Patch coalesce

webmiche opened the pull request:
The first patch for coalesce. This only applies the redundant heuristic, i.e., it checks for sets that are contained in others and deletes those.

I would especially like to get feedback on the new looping technique. Currently, I mark `FlatAffineConstraints` as being eliminated and then stop considering them.

webmiche committed:
[MLIR] add isRedundant to Simplex

webmiche committed:
[MLIR] add simple coalesce

webmiche committed:
[MLIR] add coalesce tests

webmiche committed:
[MLIR] add coalesce tests

webmiche committed:
merge coalesce into PresburgerSet

webmiche committed:
merge coalesce into PresburgerSet

webmiche commented:
For the current version, I decided to return a PresburgerSet and not change the PresburgerSet `coalesce` is called on. I think there is a discussion to be had, if we want this to happen in-place or not. As it currently was not done in-place and the test depend on this fact, I left it like that.

superty commented on mlir/include/mlir/Analysis/PresburgerSet.h:
```diff
@@ -90,6 +90,11 @@ class PresburgerSet {
   /// any of the FACs in the union are unbounded.
   bool findIntegerSample(SmallVectorImpl<int64_t> &sample);
 
+  /// eliminates FlatAffineConstraints, that are fully contained within other
```
I think this could be written better. e.g.

```quote
Simplifies the representation of the PresburgerSet.

In particular, removes all FACs which are subsets of other FACs in the union.
```

superty commented on mlir/include/mlir/Analysis/Presburger/Simplex.h:
```diff
@@ -234,6 +234,9 @@ class Simplex {
   void print(raw_ostream &os) const;
   void dump() const;
 
+  /// Check if the given constraint is redundant
```
Although it's clear from the function signature, it would be nice to specify in the doc that we refer to a new constraint rather than an existing constraint. e.g. "Check if the specified inequality already holds in the polytope."

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+      }
+
+      if (contained) {
+        marked[j] = true;
```
Here and below, you can skip the braces for single-statement ifs

webmiche replied:
Is this also usually done for single statement fors? (5 lines earlier or so)

superty replied:
Yes

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+
+  for (unsigned i = 0; i < basicSetVector.size(); ++i) {
+    if (!marked[i]) {
+      newSet.unionFACInPlace(basicSetVector[i]);
```
Skip braces for single statement ifs.

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+  assert(nSym == 0 && "symbolic sets are not yet supported by coalesce");
+  PresburgerSet newSet =
+      PresburgerSet::getEmptySet(this->getNumDims(), this->getNumSyms());
+  ArrayRef<FlatAffineConstraints> basicSetVector =
```
I don't understand why this is needed. Just use `flatAffineConstraints` directly?

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+      FlatAffineConstraints bs2 = basicSetVector[j];
+      Simplex simplex(bs2);
+
+      bool contained = true;
```
I think you can factor this out into a separate function, probably `Simplex::contains(const FlatAffineConstraints &fac)`. You can then also return early when any constraint turns out to be non-redundant. (I'm suggesting making it a member function of Simplex rather than FlatAffineConstraints so that you can re-use the same Simplex object many times as suggested above.)

webmiche replied:
Done.
I decided to name the function `Simplex::containedIn(const FlatAffineConstraints &fac)` as it actually checks whether the polytope behind the simplex is `contained in` the specified FAC.

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
 }
 
+PresburgerSet PresburgerSet::coalesce() {
+  assert(nSym == 0 && "symbolic sets are not yet supported by coalesce");
```
I think the same code should work with sets with symbols, so this shouldn't be necessary.

superty commented on mlir/unittests/Analysis/Presburger/SimplexTest.cpp:
```diff
@@ -383,4 +383,17 @@ TEST(SimplexTest, addInequality_already_redundant) {
+  EXPECT_FALSE(simplex.isRedundant({0, 1, -2}));  // y >= 2.
+}
+
 } // namespace mlir
```
Can you add a test where the new inequality doesn't make the set empty?

webmiche replied:
I was not 100% sure what you meant here. I decided to change the test polytope x >= 0, y <= 2, y = x and added the check for y >= 1, which is, in this case, a cut inequality and not redundant while not making the set empty.

superty commented on mlir/lib/Analysis/Presburger/Simplex.cpp:
```diff
@@ -1195,4 +1195,9 @@ void Simplex::print(raw_ostream &os) const {
 void Simplex::dump() const { print(llvm::errs()); }
 
+bool Simplex::isRedundant(ArrayRef<int64_t> coeffs) {
+  Optional<Fraction> minimum = computeOptimum(Direction::Down, coeffs);
```
Doesn't this assert-fail when the new inequality makes the set empty?

webmiche replied:
I am also not 100% sure what you mean here, I guess it relates to the similar comment above.

superty replied:
Nevermind, I was confused.

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+    unsigned numIneq = bs1.getNumInequalities();
+    unsigned numEq = bs1.getNumEqualities();
+    // check if bs1 is contained in any basicSet
+    for (unsigned j = 0; j < basicSetVector.size(); j++) {
```
Please use `for (unsigned j = 0, e = basicSetVector.size(); j < e; ++j) {` form.

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+      flatAffineConstraints;
+  SmallVector<bool, 4> marked(this->getNumFACs());
+
+  for (unsigned i = 0; i < basicSetVector.size(); i++) {
```
Here and below, please avoid calling a function in the loop condition unless necessary, e.g., `for (unsigned i = 0, e = basicSetVector.size(); i < e; ++i) {`. Also, use, preincrements.

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+      if (j == i || marked[j])
+        continue;
+      FlatAffineConstraints bs2 = basicSetVector[j];
+      Simplex simplex(bs2);
```
I think you should flip the loop and construct the Simplex object in the outer loop. Then you only need construct it once per basic set instead of one per pair of basic sets. (worst case)

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+  for (unsigned i = 0; i < basicSetVector.size(); i++) {
+    if (marked[i])
+      continue;
+    FlatAffineConstraints bs1 = basicSetVector[i];
```
Is this copy necessary?

superty commented on mlir/lib/Analysis/PresburgerSet.cpp:
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
+    for (unsigned j = 0; j < basicSetVector.size(); j++) {
+      if (j == i || marked[j])
+        continue;
+      FlatAffineConstraints bs2 = basicSetVector[j];
```
Is this copy necessary?

superty commented on mlir/unittests/Analysis/PresburgerSetTest.cpp:
```diff
@@ -591,4 +591,64 @@ TEST(SetTest, isEqual) {
 
+void expectCoalesce(size_t expectedNumBasicSets, PresburgerSet set) {
+  PresburgerSet newSet = set.coalesce();
+  set.dump();
```
We probably don't want to dump output for passing cases. If you want to keep this, I would suggest printing it only for failing cases.

superty reviewed (commented):
I think returning a new set is fine for now. Re: looping strategy, I left some inline comments.

webmiche committed:
move containedIn functionality to Simplex, refactor

webmiche committed:
further refactoring

superty commented on mlir/unittests/Analysis/PresburgerSetTest.cpp:
```diff
@@ -591,4 +591,64 @@ TEST(SetTest, isEqual) {
+  EXPECT_TRUE(expectedNumBasicSets == newSet.getNumFACs());
+}
+
+TEST(SetTest, coalescecontainedOneDim) {
```
Please change it to coalesceContained, coalesceSeparate

superty commented on mlir/unittests/Analysis/PresburgerSetTest.cpp:
```diff
@@ -591,4 +591,64 @@ TEST(SetTest, isEqual) {
+                         });
+  expectCoalesce(1, set);
+}
+
```
Please also add intersecting cases (some constraints are redundant but some are not) a case where one set is empty and the other is not, and case where both sets are empty.

superty replied:
And also case where one is universe, both are universe, one is universe and one is empty

webmiche replied:
Added.
You were right, when the first set is empty, the current implementation crashed. I took care of that by checking whether it is and, if so, marking it to be removed and continuing.
When both sets are empty, the current implementation returns a set without FACs in the FACvector. Is this how we represent empty sets? :)

superty replied:
Yes, the union of nothing is considered to be empty (since the identity for union is the empty set)

superty replied:
> I took care of that by checking whether it is and, if so, marking it to be removed and continuing.

The contains thing should support empty FACs. Please also add unit tests for it

webmiche replied:
Added a ton of additional tests

webmiche committed:
add test cases and support for edge cases

webmiche committed:
add tests for containedIn

webmiche committed:
add self-contained tests

webmiche committed:
add utils.h for redundant functionality

Groverkss commented on mlir/unittests/Analysis/Presburger/SimplexTest.cpp:
```diff
@@ -7,6 +7,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "mlir/Analysis/Presburger/Simplex.h"
+#include "../utils.h"
```
Can this path be made absolute?

webmiche replied:
I would have to move this into the include directory. I am not certain how these kinds of things are usually done, but if you feel having them in the includes makes more sense, I can certainly do that.

Groverkss commented on mlir/unittests/Analysis/PresburgerSetTest.cpp:
```diff
@@ -15,6 +15,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "mlir/Analysis/PresburgerSet.h"
+#include "./utils.h"
```
Can this path be made absolute?

Groverkss commented on mlir/unittests/Analysis/utils.h:
```diff
@@ -0,0 +1,41 @@
+//===- utils.h - utility functions for tests ------------------------------===//
```
I think all file names in the MLIR codebase follow UpperCamelCase. It would be better to rename this file to "FACUtils.h" or "Utils.h".

webmiche committed:
[MLIR] Introduce coalesce for PresburgerSet

This patch provides functionality for simplifying `PresburgerSet`s by checking if any `FlatAffineConstraints` in the set is contained in another, and removing such redundant FACs.

This is part of a series of patches to provide functionality for [integer set coalescing](http://impact.gforge.inria.fr/impact2015/papers/impact2015-verdoolaege.pdf) in MLIR.

Differential Revision: https://reviews.llvm.org/D110617

webmiche committed:
rename utils.h to FACUtils.h

webmiche committed:
update FACUtils license

webmiche committed:
Merge branch 'patch-coalesce' of https://github.com/Superty/llvm-project into patch-coalesce

webmiche committed:
add header guard to FACUtils.h

webmiche committed:
address review comments Pt. 1

webmiche committed:
address review comments Pt.2
